File: client/src/models/Question.ts
export interface QuestionOptionProps {
    id: number;
    text: string;
    correct?: boolean;
}

export class QuestionOption {
    public readonly id: number;
    public readonly text: string;
    public readonly correct: boolean;

    constructor(props: QuestionOptionProps) {
        this.id = props.id;
        this.text = props.text;
        this.correct = props.correct ?? false;
    }
}

export interface QuestionProps {
    id: number;
    text: string;
    answer: string;
    difficulty: string;
    options: QuestionOptionProps[];
    themeId?: string;
}

export class Question {
    public readonly id: number;
    public readonly text: string;
    public readonly answer: string;
    public readonly difficulty: string;
    public readonly options: QuestionOption[];
    public readonly themeId?: string;

    constructor(props: QuestionProps) {
        this.id = props.id;
        this.text = props.text;
        this.answer = props.answer;
        this.difficulty = props.difficulty;
        this.options = props.options.map((o) => new QuestionOption(o));
        this.themeId = props.themeId;
    }

    public getOptionById(optionId: number) {
        return this.options.find((o) => o.id === optionId);
    }
}

File: server/src/main/java/com/quest/models/Question.java
package com.quest.models;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import com.quest.enums.Difficulty;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;

@Entity
@Table(name = "questions")
public class Question {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank(message = "Question text is required")
    @Size(min = 1, max = 500, message = "Question text must be between 1 and 255")
    @Column(name = "question_text", length = 255, nullable = false)
    private String questionText;

    @ManyToOne(optional = false)
    @JoinColumn(name = "theme_id", referencedColumnName = "id", nullable = false)
    private Theme theme;

    @NotNull(message = "Difficulty is required")
    @Column(name = "difficulty", nullable = false)
    @Enumerated(EnumType.STRING)
    private Difficulty difficulty;

    @OneToMany(mappedBy = "question", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<QuestionOption> options = new ArrayList<>();

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getQuestionText() {
        return questionText;
    }

    public void setQuestionText(String questionText) {
        this.questionText = questionText;
    }

    public Difficulty getDifficulty() {
        return difficulty;
    }

    public void setDifficulty(Difficulty difficulty) {
        this.difficulty = difficulty;
    }

    public List<QuestionOption> getOptions() {
        return options;
    }

    public void setOptions(List<QuestionOption> options) {
        this.options = options;
    }

    public Theme getTheme() {
        return theme;
    }

    public void setTheme(Theme theme) {
        this.theme = theme;
    }

    public Optional<QuestionOption> getOptionById(Long optionId) {
        return options.stream()
                .filter(option -> option.getId().equals(optionId))
                .findFirst();
    }
}
